### ISSUE TITLE
Modifying User.companies results in AssertionError

### ISSUE THREAD

**Issue by fredrikaverpil** [eoyilmaz/stalker#27]
I'm creating a user with some companies:

```python
c1 = Client(name='Company X')
c2 = Client(name='Company Y')
c3 = Client(name='Company Z')

user = User(
		name='Fredrik',
		login='fredrik',
		email='f@f.f',
		password='pass',
		companies=[c1, c2, c3]
		) 

db.DBSession.add( user )
db.DBSession.commit()
```
Then I modify the companies, effectively removing the old ones:
```python
c1 = Client(name='New Company X')
c3 = Client(name='New Company Y')
c2 = Client(name='New Company Z')

user = User.query.filter_by(name='Fredrik').first()
user.companies = [c1, c2, c3]

db.DBSession.commit() # <--- triggers the error
```

And this causes an error:
```python
AssertionError: Dependency rule tried to blank-out primary key column 'Client_Users.uid' on instance '<ClientUser at 0x58f60f0>'
```

Do you see this on your end too?

**Comment by eoyilmaz** [comment]
The following is working for me without a problem:

    new_companies = []
    c1 = Client(name='Company X')
    c2 = Client(name='Company Y')
    db.DBSession.add_all([c1, c2])
    db.DBSession.commit()

    c1 = Client.query.filter_by(name='Company X').first()
    c2 = Client.query.filter_by(name='Company Y').first()

    new_companies.append(c1)
    new_companies.append(c2)
    user = User(name='test_user', password='1234', email='a@a.com',
                login='test_user')
    user.companies = new_companies
    db.DBSession.commit()

    assert c1 in user.companies
    assert c2 in user.companies

**Comment by eoyilmaz** [comment]
The following also is working for me:

    new_companies = []
    c1 = Client(name='Company X')
    c2 = Client(name='Company Y')
    c3 = Client(name='Company Z')
    db.DBSession.add_all([c1, c2, c3])
    db.DBSession.commit()

    c1 = Client.query.filter_by(name='Company X').first()
    c2 = Client.query.filter_by(name='Company Y').first()

    new_companies.append(c1)
    new_companies.append(c2)
    user = User(name='test_user', password='1234', email='a@a.com',
                login='test_user', clients=[c3])
    db.DBSession.commit()

    user.companies = new_companies
    db.DBSession.commit()

    self.assertTrue(c1 in user.companies)
    self.assertTrue(c2 in user.companies)
    self.assertTrue(c3 not in user.companies)

**Comment by fredrikaverpil** [comment]
I revised the code in my initial post. Please see if that makes sense to you.

**Comment by eoyilmaz** [comment]
also this one is working:

    new_companies = []
    c1 = Client(name='Company X')
    c2 = Client(name='Company Y')
    c3 = Client(name='Company Z')
    db.DBSession.add_all([c1, c2, c3])
    db.DBSession.commit()

    c1 = Client.query.filter_by(name='Company X').first()
    c2 = Client.query.filter_by(name='Company Y').first()
    c3 = Client.query.filter_by(name='Company Z').first()

    user = User(name='test_user', password='1234', email='a@a.com',
                login='test_user', clients=[c3])
    db.DBSession.commit()

    new_companies.append(c1)
    new_companies.append(c2)
    user.companies = new_companies
    db.DBSession.commit()

    assert c1 in user.companies
    assert c2 in user.companies
    assert c3 not in user.companies

**Comment by eoyilmaz** [comment]
Yes I confirm that causes an error, I'm investigating the cause.

Erkan Ozgur Yilmaz

**Comment by fredrikaverpil** [comment]
> Yes I confirm that causes an error, I'm investigating the cause.

It seems that if you add the User to the database in the same run as when you modify its companies, it works, like in your code examples. But if the user already exists you get this error.

**Comment by eoyilmaz** [comment]
It seems that I forgot to add the cascade rule to the ``User.company_role`` relation, adding ``cascade='all, delete-orphan'`` fixed the error.

Fixed in 84f3cbe217ee96da4ee070bde8ae6617a0dae353
